cartographer/pose_graph/pose_graph_data.cc

```diff
@@ -0,0 +1,96 @@
+ */
+
+#include "cartographer/pose_graph/pose_graph_data.h"
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/pose_graph/constraint/acceleration_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/interpolated_relative_pose_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/relative_pose_constraint_2d.h"
+#include "cartographer/pose_graph/constraint/relative_pose_constraint_3d.h"
+#include "cartographer/pose_graph/constraint/rotation_constraint_3d.h"
+
+namespace cartographer {
+namespace pose_graph {
+namespace {
+
+std::unique_ptr<Constraint> CreateConstraint(
+    const proto::Constraint& constraint) {
+  const auto& id = constraint.id();
+  const auto& loss = constraint.loss_function();
+  const auto& cost = constraint.cost_function();
+  switch (cost.type_case()) {
+    case (proto::CostFunction::kRelativePose2D):
+      return common::make_unique<RelativePoseConstraint2D>(
+          id, loss, cost.relative_pose_2d());
+    case (proto::CostFunction::kRelativePose3D):
+      return common::make_unique<RelativePoseConstraint3D>(
+          id, loss, cost.relative_pose_3d());
+    case (proto::CostFunction::kAcceleration3D):
+      return common::make_unique<AccelerationConstraint3D>(
+          id, loss, cost.acceleration_3d());
+    case (proto::CostFunction::kRotation3D):
+      return common::make_unique<RotationContraint3D>(id, loss,
+                                                      cost.rotation_3d());
+    case (proto::CostFunction::kRelativePose2DInterpolated):
+      LOG(FATAL) << "Not yet implemented.";
+      return nullptr;
+    case (proto::CostFunction::kInterpolatedRelativePose3D):
+      return common::make_unique<InterpolatedRelativePoseConstraint3D>(
+          id, loss, cost.interpolated_relative_pose_3d());
+    case (proto::CostFunction::TYPE_NOT_SET):
+      LOG(FATAL) << "Constraint cost function type is not set.";
+  }
+  return nullptr;
+}
+
+}  // namespace
+
+void AddNodeToPoseGraphData(const proto::Node& node, PoseGraphData* data) {
+  NodeId node_id(node.id());
+  switch (node.parameters().type_case()) {
+    case (proto::Parameters::kPose2D): {
+      data->nodes.pose_2d_nodes.emplace(
+          node_id,
+          Pose2D(node_id, node.constant(), node.parameters().pose_2d()));
+      return;
+    }
+    case (proto::Parameters::kPose3D): {
+      data->nodes.pose_3d_nodes.emplace(
+          node_id,
+          Pose3D(node_id, node.constant(), node.parameters().pose_3d()));
+      return;
+    }
+    case (proto::Parameters::kImuCalibration): {
+      data->nodes.imu_calibration_nodes.emplace(
+          node_id, ImuCalibration(node_id, node.constant(),
+                                  node.parameters().imu_calibration()));
+      return;
+    }
+    case (proto::Parameters::TYPE_NOT_SET): {
+      LOG(FATAL) << "Node parameter type is not set.";
+      return;
+    }
+  }
+}
+
+void AddConstraintToPoseGraphData(const proto::Constraint& constraint,
```

Comment:
How about this:
Expose the CreateConstraint/CreateNode functions instead and fill the vector/map in the caller. I think this is more flexible than the current interface.

Reply:
I am not sure that I understand how to do that with nodes.

Reply:
Yes, let's keep as is.